Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h

```diff
@@ -27,7 +27,32 @@ functionality for accessing the data stream and processing data.
 */
 class DLLExport IKafkaStreamDecoder {
 public:
-  using CallbackFn = std::function<void()>;
+  /**
+   * Defines a thread-safe callback. A mutex is held
+   * both during assignment of a new callback function
+   * and during the call.
+   */
+  class Callback {
+  public:
+    using FnType = std::function<void()>;
+
+    Callback(Callback::FnType callback) : m_mutex(), m_callback() {
+      setFunction(std::move(callback));
```

Comment:
I believe constructors are the exception.

https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html